feat: avm/res/network/load balancer create public IP Address or public IP prefix

arnoldna opened the pull request:
## Description

- Added functionality to support the creation of public IP addresses and public IP prefixes in load balancer `frontendIPconfigurations`. This new functionality utilizes the AVM modules for `public-ip-address` and `public-ip-prefix`
- Introduced new test cases for public IP address and public IP prefix creation in the load balancer module.

closes #2452 

## Pipeline Reference

| Pipeline |
| -------- |
[![avm.res.network.load-balancer](https://github.com/arnoldna/bicep-registry-modules/actions/workflows/avm.res.network.load-balancer.yml/badge.svg?branch=avm%2Fres%2Fnetwork%2Fload-balancer-pip)](https://github.com/arnoldna/bicep-registry-modules/actions/workflows/avm.res.network.load-balancer.yml)

## Type of Change

<!-- Use the checkboxes [x] on the options that are relevant. -->

- Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
  - [X] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [X] Update to documentation
- [ ] Update to CI Environment or utilities (Non-module affecting changes)

## Checklist

- [ ] I'm sure there are no other open Pull Requests for the same update/change
- [X] I have run `Set-AVMModule` locally to generate the supporting module files.
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings
- [ ] I have updated the module's CHANGELOG.md file with an entry for the next version

<!--  Please keep up to date with the contribution guide at https://aka.ms/avm/contribute/bicep -->

arnoldna committed:
Add Bicep tests for external public IP and public IP prefix creation

- Implemented `main.test.bicep` for testing external public IP load balancer with required parameters.
- Created `main.test.bicep` for testing external public IP prefix load balancer with required parameters.
- Both tests include diagnostics module and resource group creation.
- Configured frontend IP settings for public IP and public IP prefix with appropriate SKU and zones.

arnoldna committed:
Merge branch 'main' of https://github.com/arnoldna/bicep-registry-modules into avm/res/network/load-balancer-pip

arnoldna committed:
AVM-RES: Remove enableTelemetry from public IP configurations

- Removed enableTelemetry parameter from loadBalancer_publicIPAddresses and loadBalancer_publicIPPrefixes modules.
- This change simplifies the configuration by relying on default telemetry settings.

arnoldna committed:
AVM-RES: Add telemetry parameters for referenced modules

- Introduced 'enableReferencedModulesTelemetry' parameter to control telemetry for referenced modules.
- Updated load balancer modules to utilize the new telemetry parameter.

arnoldna committed:
AVM-RES: Update telemetry handling in load balancer module

- Remove enableReferencedModulesTelemetry parameter from the load balancer module.
- Set enableReferencedModulesTelemetry variable to false.
- Update telemetry logic in public IP address and prefix modules to use the new variable.

AlexanderSehr commented on avm/res/network/load-balancer/main.bicep:
```diff
@@ -85,11 +99,23 @@ var frontendIPConfigurationsVar = [
+              })
+        : (contains(frontendIPConfiguration, 'publicIPPrefix') && !empty(frontendIPConfiguration.publicIPPrefix)
+            ? {
+                id: frontendIPConfiguration.publicIPPrefix
```
Note: This would be a breaking change

AlexanderSehr commented on avm/res/network/load-balancer/main.bicep:
```diff
@@ -68,11 +70,23 @@ var frontendIPConfigurationsVar = [
+            : {
+                id: resourceId(
+                  'Microsoft.Network/publicIPAddresses',
+                  frontendIPConfiguration.?pipConfiguration.?name ?? '${name}-pip-${index}'
```
Could potentially be replaced with an output from the deployment by referencing the resource via the index of the loop

AlexanderSehr commented on avm/res/network/load-balancer/tests/e2e/external-pip-creation/main.test.bicep:
```diff
@@ -0,0 +1,92 @@
+    name: '${uniqueString(deployment().name, resourceLocation)}-test-${serviceShort}-${iteration}'
+    params: {
+      name: '${namePrefix}${serviceShort}001'
+      location: resourceLocation
```
```suggestion
```
Redundant

arnoldna committed:
AVM-RES: Update load balancer module and tests

- Removed zones and location parameters from load balancer configurations in README and test files.
- Updated managed identity resource API version in multiple test dependencies.
- Added new test for external public IP prefix creation.
- Deleted obsolete external public IP prefix test file.

arnoldna commented on avm/res/network/load-balancer/tests/e2e/external-pip-creation/main.test.bicep:
```diff
@@ -0,0 +1,92 @@
+            skuName: 'Standard'
+            skuTier: 'Regional'
+            allocationMethod: 'Static'
+            zones: [
```
It is using the latest version of the module. I was able to remove the zones parameter from this test and the deployment reflects all three availability zones.

AlexanderSehr commented on avm/res/network/load-balancer/tests/e2e/external-pip-creation/main.test.bicep:
```diff
@@ -18,26 +18,12 @@ param resourceLocation string = deployment().location
 
 @description('Optional. A token to inject into the name of each resource.')
-param namePrefix string = '#_namePrefix_#'
+param namePrefix string = 'nga'
```
Beware 😄 

```suggestion
param namePrefix string = '#_namePrefix_#'
```

arnoldna committed:
feat: Enhance Load Balancer module with public IP and prefix configurations

- Added definitions for publicIPAddressConfigurationType and publicIPPrefixConfigurationType in main.json.
- Updated frontendIPConfigurationType to support new public IP and prefix configurations.
- Modified existing parameters and properties to align with new configurations.
- Updated test cases to reflect changes in publicIPAddressId to publicIPAddressResourceId.
- Adjusted test cases for external and internal configurations to use new naming conventions.
- Bumped version to 0.6 in version.json to reflect the new features and changes.

arnoldna committed:
Adjusted logic in main.bicep and main.json to handle nullable prefix length.

arnoldna committed:
AVM-RES: Refactor public IP and prefix handling in load balancer

- Simplified logic for assigning public IP addresses and prefixes in main.bicep and main.json.
- Enhanced readability and maintainability of the code.
- Ensured compatibility with existing configurations for frontend IPs.

arnoldna committed:
AVM-RES: Update load balancer zones logic

- Refactored the logic for determining availability zones in the load balancer configuration.
- Changed the condition to use `contains` for checking availability zones.
- Updated the JSON template hash for consistency.

arnoldna committed:
AVM-RES: Update load balancer configuration parameters

- Changed 'zones' to 'availabilityZones' in load balancer configurations.
- Updated 'subnetId' to 'subnetResourceId' for consistency in parameter naming.
- Adjusted test deployment to reflect changes in frontend IP configurations.

arnoldna committed:
Update avm/res/network/load-balancer/CHANGELOG.md

Co-authored-by: Alexander Sehr <ASehr@hotmail.de>

arnoldna committed:
Merge branch 'main' into avm/res/network/load-balancer-pip

arnoldna committed:
AVM-RES: Update changelog formatting for parameter names

- Adjusted formatting of parameter name changes in the changelog for clarity.
- Ensured compliance with AVM naming standards for resource IDs.

arnoldna committed:
Merge branch 'main' into avm/res/network/load-balancer-pip
